Clean Powerline; Add Dir Support to Clean_Files

davidpfarrell opened the pull request:
##  Adds full-directory support for clean_files.txt

_lint_clean_files.sh_
```
# Pull list of checkable files from clean_files.txt
#  - Folder references are allowed
#  - Empty lines are ignored
#  - Comment lines (#) are ignored
mapfile -t FILES < <(
	cat clean_files.txt \
		| grep -v -E '^\s*$' \
		| grep -v -E '^\s*#' \
		| xargs -n1 -I{} find "{}" -type f
)

BASH_IT='' pre-commit run --files "${FILES[@]}"
```

## Cleans Primary Powerline Theme
To test the directory support, we add `themes/powerline` to the file.
This PR incorporates all the auto-fixes and recommended changes resulting from executing the pre-commit command locally:
```
BASH_IT='' pre-commit run --files $(cat clean_files.txt | xargs -n1 -I{} find {} -type f);
```

NOTE: Using the `"Hide whitespace changes"` features on the `"Files Changed"` tab will make the PR much easier to review.

[edit] Updated local pre-commit run to include prefix `BASH_IT=''` which is needed to suss out source includes that need a `"# shellcheck source"` directive.

davidpfarrell commented on themes/powerline/powerline.base.bash:
```diff
@@ -1,282 +1,280 @@
-  LAST_SEGMENT_COLOR=${params[1]}
-  (( SEGMENTS_AT_LEFT += 1 ))
+function __powerline_left_segment() {
+	local params
+	IFS="|" read -ra params <<< "${1}"
```
This is the most complicated change in the PR.
TIL about the `read` function.
Testing locally confirmed that, if you create the `local` variable first, then the read operation will populate the existing variable (vs creating a new, global variable)

davidpfarrell commented:
TIL that when running the pre-commit checks locally, its best to clear out the `BASH_IT` variable to suss out source includes that need a `"# shellcheck source"` directive.

```
BASH_IT='' pre-commit run --files $(cat clean_files.txt | xargs -n1 -I{} find {} -type f);
```

nwinkler commented on clean_files.txt:
```diff
@@ -1 +1,2 @@
 themes/90210/90210.theme.bash
```
Maybe change this to the directory name as well? Remove the file name?

davidpfarrell replied:
Yeah good idea plus its an easy win for this theme since it only contains the one file anyways.

nwinkler reviewed (commented):
One minor comment/question, please see inline.

Thanks for doing this!

Does it make sense to add the `BASH_IT=''` to the documentation to show how to run `pre-commit` locally? Since the command is getting quite long with all the file/directory parsing, wouldn't it be good to add a script to run `pre-commit` locally?

davidpfarrell commented on docs/make.bat:
```diff
@@ -1,35 +1,35 @@
-
-:end
-popd
+@ECHO OFF
+
+pushd %~dp0
+
+REM Command file for Sphinx documentation
+
+if "%SPHINXBUILD%" == "" (
+	set SPHINXBUILD=sphinx-build
+)
+set SOURCEDIR=.
+set BUILDDIR=_build
+
+if "%1" == "" goto help
+
+%SPHINXBUILD% >NUL 2>NUL
+if errorlevel 9009 (
+	echo.
+	echo.The 'sphinx-build' command was not found. Make sure you have Sphinx
+	echo.installed, then set the SPHINXBUILD environment variable to point
+	echo.to the full path of the 'sphinx-build' executable. Alternatively you
+	echo.may add the Sphinx directory to PATH.
+	echo.
+	echo.If you don't have Sphinx installed, grab it from
+	echo.http://sphinx-doc.org/
+	exit /b 1
+)
+
+%SPHINXBUILD% -M %1 %SOURCEDIR% %BUILDDIR% %SPHINXOPTS% %O%
+goto end
+
+:help
+%SPHINXBUILD% -M help %SOURCEDIR% %BUILDDIR% %SPHINXOPTS% %O%
+
+:end
+popd
```
I got confused for a bit, but I *think* what is happening here is that the line-endings for the respository-version of the file were changed after I  : 
* Added ".bat" extension to .gitattributes:
```
*.bat text eol=crlf
```
* Normalized the file:
```
git add --renormalize docs/make.bat 
```

If you hide whitespace changes then this file will hide, but other than that I 'm  not sure how to confirm if the *correct* version of the file will end up in the repo.

davidpfarrell commented:
> Does it make sense to add the `BASH_IT=''` to the documentation to show how to run `pre-commit` locally? Since the command is getting quite long with all the file/directory parsing, wouldn't it be good to add a script to run `pre-commit` locally?

For running against the clean_files list, I moved the logic into a script, `lint_clean_files.sh` (open to better names) and also modified the travis file to use it.

There's not a great way to unset the BASH_IT variable as par of git's standard per-commit invocation.

But maybe a small section in the docs for using `BASH_IT='' pre-commit` could be useful for testing local changes before committing.

NoahGorny commented:
Wow @davidpfarrell , great work!
I like the idea of the script and the changes you made, and good job on fixing the broken file. I think it got in as my branch was not rebased when I merged the docs PR.

I think those changes should be in a different PR than this. We should merge the changes and then this PR.

About the docs- please put the sections in `docs/development.rst` and not in `development.md` which will soon be deleted. You can also view the docs in https://bash-it.readthedocs.io/en/latest/

I will give you guys maintainer status in readthedocs.org so you will be able to manage the docs as well

davidpfarrell commented:
> I think those changes should be in a different PR than this. We should merge the changes and then this PR.

Indeed this PR got away from me and contain(ed) more disparate changes that I like.

As a result, I have moved some of the changes into 3 other PRs:

* Adds better support for .bat files; Fixes docs/make.bat (#1700) (MERGED)
* Moves lint logic into lint_clean_files.sh (#1701) (NEEDS TO BE MERGED BEFORE THIS ONE)
* Adds manual pre-commit usage to docs/development.rst (#1702) (NOT A BLOCKER)

With those PRs moved out, these PR reduces to 5 files, and once #1701 is merged, this PR will reduce to 4 files.

NoahGorny commented:
Please rebase your PR on master in order to allow us to see the correct diff
I also see you added and extended the scripts here, I will look after you will update this PR

davidpfarrell commented:
> Please rebase your PR on master in order to allow us to see the correct diff
> I also see you added and extended the scripts here, I will look after you will update this PR

Hi @NoahGorny ,

I merged master, resolved conflicts, then pushed, but then decided to squash since the original commit history was so messy.

So now there's only one commit to be merged, even though the actual PR diff hasn't changed.

Thanks and lemme know if you have an questions/observations!

NoahGorny commented on lint_clean_files.sh:
```diff
@@ -1,7 +1,16 @@
+)
 
-pre-commit run --files "${FILES[@]}"
+BASH_IT='' pre-commit run --files "${FILES[@]}"
```
You should add a comment here explaining why you do this

davidpfarrell replied:
Done, good suggestion, thanks !

NoahGorny reviewed (approved):
Just minor nitpick, change it and it's a go from me!

NoahGorny commented:
Great job @davidpfarrell !
